metals/src/main/scala/scala/meta/internal/worksheets/WorksheetProvider.scala

```diff
@@ -108,8 +110,20 @@ class WorksheetProvider(
       path: AbsolutePath,
       token: CancelToken
   ): Future[Unit] = {
-    evaluateAsync(path, token).map(
-      _.foreach(publisher.publish(languageClient, path, _))
+    val possibleBuildTarget = buildTargets.inverseSources(path)
+    val previouslyCompiled = compilations.previouslyCompiled.toSeq
+
+    (possibleBuildTarget match {
+      case Some(bdi) if previouslyCompiled.isEmpty =>
+        compilations.compileTarget(bdi).ignoreValue
+      case Some(bdi) if !previouslyCompiled.contains(bdi) =>
+        compilations.compileTarget(bdi).ignoreValue
+      case _ =>
+        Future.successful(())
+    }).flatMap(_ =>
```

Comment:
There is still one flaw with that that I'm unsure to get around, _however_ I think it's related to some of the multiple other tickets we have about improving the standalone presentation compiler design. The main problem is that when you directly open a worksheet that is located in a src directory, it may not recognize the build target before triggering the evaluation. So this check here shows the build target as empty, then triggers an evaluation even though it really does belong to a build target. However, then as soon as it connects to the build server, then this ends up being hit again, it does recognize that the worksheet belongs to a build target, then correctly compiles the target, and corrects the worksheet evaluation. So when playing around with this, you _may_ see for a moment that the evaluation is incorrect since it thinks there is no build target, and therefore doesn't recognize your import, but then it quickly corrects.

Reply:
Yeah, that's exactly the same issue as in other places. Probably we should wait for the indexing promise in this case.

Reply:
I will work on this before the next release

Reply:
Yea that's what I figured. For the time being, do you think this change is fine though? I do think it's a significant improvement on the current situation, and when testing it, you don't really hit on this issue at all unless you are open your worksheet as the first document in a large build. 

So most of the time, it should be fine, like so:

![2020-10-15 12 45 18](https://user-images.githubusercontent.com/13974112/96113582-a9db8380-0ee4-11eb-98bd-d99d3765e39c.gif)

Reply:
I think it's fine for now.